[Caracal] Adapt Tox and Github Workflow &  [Caracal] Fix Unit Tests

sven-rosenzweig opened the pull request:
[Caracal] Adapt Tox and Github Workflow
Adapt github workflow python version and branch to the Caracal/2024.1 release.

[Caracal] Fix Unit Tests
With the Caracal release, default policies are activated by default causing our tests to fail error 'PolicyNotAuthorized'. Simply running them as admin fixes the issue.

sven-rosenzweig committed:
[Caracal] Adapt Tox and Github Workflow

Adapt github workflow python version and branch to the Caracal/2024.1
release.

With later versions of pecan, WebTest is not a dependency of it anymore,
so in order to continue using this, an additional import in
test-requirements is required.

sven-rosenzweig committed:
[Caracal] Fix Unit Tests

With the Caracal release, default policies are activated by default
causing our tests to fail error 'PolicyNotAuthorized'.
Simply running them as admin fixes the issue.

sebageek commented on asr1k_neutron_l3/tests/unit/plugins/l3/service_plugins/test_l3_extension_adapter.py:
```diff
@@ -326,11 +327,13 @@ def test_router_create_with_extended_nat_pool_remove_gateway(self, pc_mock):
 
                 with mock.patch.object(ASR1KPluginBase, 'ensure_default_route_skip_monitoring', autospec=True):
-                    upd_router = self._remove_external_gateway_from_router(router['router']['id'], None)
+                    # upd_router = self._remove_external_gateway_from_router(router['router']['id'], None)
```
Do we need the commted out call? I think we can remove it, right?

sebageek commented on asr1k_neutron_l3/plugins/db/asr1k_db.py:
```diff
@@ -684,25 +713,27 @@ def __init__(self, context, router_id, port, segment):
-        entry = self.session.query(asr1k_models.ASR1KExtraAttsModel).filter_by(router_id=self.router_id,
+        with db_api.CONTEXT_READER.using(self.context):
+            entry = self.context.query(asr1k_models.ASR1KExtraAttsModel).filter_by(router_id=self.router_id,
                                                                                agent_host=self.agent_host,
                                                                                port_id=self.port_id,
                                                                                segment_id=self.segment_id
                                                                                ).first()
```
Could be indented to the same level.

sebageek commented on asr1k_neutron_l3/plugins/db/asr1k_db.py:
```diff
@@ -684,25 +713,27 @@ def __init__(self, context, router_id, port, segment):
                                                                                port_id=self.port_id,
                                                                                segment_id=self.segment_id
                                                                                ).first()
         return entry is not None
```
Should be indented into the `with`, I think.

sebageek commented on asr1k_neutron_l3/plugins/db/asr1k_db.py:
```diff
@@ -684,25 +713,27 @@ def __init__(self, context, router_id, port, segment):
-        if len(second_dot1qs_available) == 0:
-            raise asr1k_exceptions.SecondDot1QPoolExhausted(agent_host=self.agent_host)
-        self.second_dot1q = random.choice(second_dot1qs_available)
+        with db_api.CONTEXT_READER.using(self.context):
```
Could be a decorator.

sven-rosenzweig replied:
Answer can be found [here](https://github.com/sapcc/asr1k-neutron-l3/pull/120/commits/41a0a1032febbd1c0212c77a985935387afc844e#r1862248940). 

I think we need a context parameter to use the decorator.

sebageek commented on asr1k_neutron_l3/plugins/l3/service_plugins/l3_extension_adapter.py:
```diff
@@ -208,29 +208,28 @@ def get_hosts_for_network(self, context, network_id):
-                                                                                      ).first()
-                    if entry:
-                        entry.update(extra_att)
+        with db_api.CONTEXT_WRITER.using(context):
```
Could be a decorator.

sven-rosenzweig replied:
to make this work with a decorate the contest must be passed as parameter 
```
 @db_api.CONTEXT_READER
    def set_next_entries(self, context):
        extra_atts = context.query(asr1k_models.ASR1KExtraAttsModel).filter_by(agent_host=self.agent_host)
```

sven-rosenzweig replied:
my suggestion is to leave it as is

sven-rosenzweig committed:
[GH Workflow] Run TOX version 4

PEP-660 introduced a standardized way of installing a package in development mode,
providing the same effect as if pip install -e was used.

If use_develop (defaults to false) is set, tox install the current
package in development mode using PEP 660.

With this setting enabled, our unittests fail with 'No 'neutron.service_plugins'
driver found, looking for 'asr1k_l3_routing''.

Using 'pip install -e.' fixes this.
